enhancements/cgu-precaching-additional-images.md

```diff
@@ -0,0 +1,308 @@
+- N/A
+
+superseded-by:
+- N/A
+---
+
+# Enable pre-caching of user-specified images
+
+## Summary
+
+The existing pre-caching feature on the Topology Aware Lifecycle Manager (TALM) allow users to
+precache OpenShift specific workloads on managed clusters. This enhancement proposal (EP) suggests
+adding a new feature to TALM to allow users to pre-cache additional application-specific workload
+images on managed clusters.
+
+
+## Motivation
+
+The bandwidth between typical SNO sites and the centralized registry(es) is limited:
+- Less than 100 Mbps in some cases
+- Round trip packet latency may also be high
+
+Furthermore, aggregate application workload image sizes to be transferred during an upgrade can be high.
+
+Customers would like to pre-cache additional workload images on the SNO sites prior to upgrading their
+application(s). This reduces the impact of slow networks and removes risk due to transient networking
+events during the upgrade.
+
+
+### User Stories
+
+- As a user, I want to pre-cache my application images so that I can reliably and efficiently upgrade
+my clusters.
+
+### Goals
+
+The user must be able to do the following:
+- Declaratively define a set of images to be pre-cached on managed SNO nodes.
+- Enable pre-caching of that set of images to one or more nodes at a time of their choosing (e.g.
+maintenance window).
+- Retrieve status indicating success or failure of the pre-cache operation.
+
+The proposed solution should not increase steady-state CPU use of cluster outside of the precaching
+operation.
+
+
+### Non-Goals
+
+The pre-caching and backup functionality of TALM is currently enabled to work on SNOs only. There is
+an item on the enhancement list to extend this functionality to operate on multi-node architectures
+(SNO+workers). This EP will also target SNOs only and the capability for multi-node configurations
+will be scoped as a future enhancement.
+
+The following items are out of scope for this EP:
+- There will be no image registry or UI for inspecting the cached images.
+- Management of the user application is not included. This epic comprises only pre-caching
+user-specified images in CRIO.
+
+
+## Proposal
+
+The proposed solution is to leverage the existing pre-caching job on TALM to pre-cache the additional
+user-specified images. This job is currently used to pre-cache the OpenShift platform image and operators
+on the managed clusters. The job runs on the clusters and pulls the requested set of missing images/layers.
+
+This proposal further allows the application of the existing Cluster Group Upgrade (GGU) custom resource (CR)
+to enable the user to pre-cache the additional images.
+
+The following changes are suggested:
+
+1. Add a new `AdditionalImages` field to the `PrecachingSpec` struct in
+`cluster-group-upgrades-operator/api/v1alpha1/clustergroupupgrade_types.go` as shown below.
+    ```golang
+    ...
+    // PrecachingSpec defines the pre-caching software spec derived from policies
```

Comment:
It isn't clear from the text here, but I think that this change only affects the status section of the CGU, is that right?

Reply:
That is correct.

Reply:
This is now changed. The proposed approach is to use a new pre-caching CR and reference it in the CGU.